packages/plugins/src/hardhat/simulate.ts

```diff
@@ -511,6 +457,124 @@ export const handleSimulationSuccess = async (
   return
 }
 
+/**
+ * Create a Proxy that wraps a `HardhatEthersProvider` to implement retry and timeout logic, which
+ * isn't available natively in this provider.
+ */
+export const createHardhatEthersProviderProxy = (
+  ethersProvider: HardhatEthersProvider
+): HardhatEthersProvider => {
+  return new Proxy(ethersProvider, {
+    get: (target, prop) => {
+      return (...args: any[]) => {
+        // Queue a snapshot of the Hardhat node state. We may revert to this snapshot later in this
+        // function to prevent a 'nonce too low' bug in Hardhat. We must queue the snapshot before
+        // calling `target[prop](...args)` to avoid this nonce bug, which occurs in spite of the
+        // fact that we don't immediately await either of these calls. We don't await this call yet
+        // because the `target[prop](...args)` call may be synchronous. If we await this call, then
+        // every call routed through this Proxy will become asynchronous, which could lead to
+        // expected behavior.
+        //
+        // More info on the nonce error is in this pull request description:
+        // https://github.com/sphinx-labs/sphinx/pull/1565.
+        const snapshotIdPromise = target.send('evm_snapshot', [])
+
+        // Call the method on the EthersJS provider.
+        const initialResult = target[prop](...args)
+        // Check if the call is synchronous. If so, return it.
+        if (!(initialResult instanceof Promise)) {
+          return initialResult
+        }
```

Comment:
You may wonder if returning early before awaiting the `snapshotIdPromise` is problematic. The answer is that I don't know. I'm going to make a ticket to investigate it. I don't think it's a big enough concern to block this PR